### PR TITLE
Add 3 LLVM patches to fix issue #19792

### LINKED ISSUE #19769: llvm-3.9.0_threads.patch breaks the build on FreeBSD
Our patched LLVM 3.7 built fine on FreeBSD, but it seems that LLVM 3.9 refuses to build on FreeBSD when llvm-3.9.0_threads.patch is applied. The following message is displayed, along with several other similar messages referencing other LLVM files:
```
CMake Error at cmake/modules/LLVMProcessSources.cmake:83 (message):
  Found unknown source file
  /usr/home/alex/repos/julia/deps/srccache/llvm-3.9.1/tools/lli/OrcLazyJIT.cpp


  Please update
  /usr/home/alex/repos/julia/deps/srccache/llvm-3.9.1/tools/lli/CMakeLists.txt
```
The build works fine if the application of this patch is commented out.

This is with CMake 3.7.1. I guess the patch was copied from Rust's fork of LLVM, so this issue may not be unique to Julia.

### LINKED ISSUE #19792: LLVM 3.9 miss-compiles inlined broadcast for Array & SubArray
On
```julia
julia> versioninfo()
Julia Version 0.6.0-dev.1807
Commit 26c8d85* (2016-12-31 04:12 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin15.6.0)
  CPU: Intel(R) Core(TM) i7-3520M CPU @ 2.90GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
```
The `linalg/qr` tests fail. The failure reduces to
```julia
n = 10;
srand(1234321);
a = randn(n,n)/2;

av = view(a, 1:n - 1, 1:n - 1);
qrav = qrfact(av);
q, r  = qrav[:Q], qrav[:R];
q*r; # fine
av; # fine
(q*r, av); # fine
q*r ≈ av # segfaults
# q*r = av # also segfaults
# q*r - av # also segfaults
```
e.g.
```julia
julia> n = 10;

julia> srand(1234321);

julia> a = randn(n,n)/2;

julia> av = view(a, 1:n - 1, 1:n - 1);

julia> qrav = qrfact(av);

julia> q, r  = qrav[:Q], qrav[:R];

julia> q*r; # fine

julia> av; # fine

julia> (q*r, av); # fine

julia> q*r ≈ av # segfaults

signal (11): Segmentation fault: 11
while loading no file, in expression starting on line 0
_ZN12_GLOBAL__N_113InlineSpiller16postOptimizationEv at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
_ZN4llvm12RegAllocBase16postOptimizationEv at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
_ZN12_GLOBAL__N_18RAGreedy20runOnMachineFunctionERN4llvm15MachineFunctionE at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
_ZN4llvm19MachineFunctionPass13runOnFunctionERNS_8FunctionE at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE at /Users/sacha/pkg/julia/usr/lib//libLLVM.dylib (unknown line)
operator() at /Users/sacha/pkg/julia/src/jitlayers.cpp:436 [inlined]
__invoke<(lambda at /Users/sacha/pkg/julia/src/jitlayers.cpp:434:13) &, llvm::Module &> at /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/__functional_base:416 [inlined]
__call<(lambda at /Users/sacha/pkg/julia/src/jitlayers.cpp:434:13) &, llvm::Module &> at /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/__functional_base:437 [inlined]
operator() at /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1437
operator() at /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1817 [inlined]
addModuleSet<llvm::SmallVector<std::__1::unique_ptr<llvm::Module, std::__1::default_delete<llvm::Module> >, 1>, llvm::RTDyldMemoryManager *, std::__1::unique_ptr<llvm::orc::LambdaResolver<(lambda at /Users/sacha/pkg/julia/src/jitlayers.cpp:530:23), (lambda at /Users/sacha/pkg/julia/src/jitlayers.cpp:554:23)>, std::__1::default_delete<llvm::orc::LambdaResolver<(lambda at /Users/sacha/pkg/julia/src/jitlayers.cpp:530:23), (lambda at /Users/sacha/pkg/julia/src/jitlayers.cpp:554:23)> > > > at /Users/sacha/pkg/julia/usr/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:73 [inlined]
addModule at /Users/sacha/pkg/julia/src/jitlayers.cpp:558
jl_add_to_ee at /Users/sacha/pkg/julia/src/jitlayers.cpp:782 [inlined]
jl_finalize_function at /Users/sacha/pkg/julia/src/jitlayers.cpp:793
getAddressForFunction at /Users/sacha/pkg/julia/src/codegen.cpp:1070 [inlined]
jl_generate_fptr at /Users/sacha/pkg/julia/src/codegen.cpp:1193
jl_call_method_internal at /Users/sacha/pkg/julia/src/./julia_internal.h:240
jl_apply_generic at /Users/sacha/pkg/julia/src/gf.c:2196
isapprox at ./linalg/generic.jl:1190
unknown function (ip: 0x3198d4456)
jl_apply_generic at /Users/sacha/pkg/julia/src/gf.c:2196
do_call at /Users/sacha/pkg/julia/src/interpreter.c:75
eval at /Users/sacha/pkg/julia/src/interpreter.c:214
jl_interpret_toplevel_expr at /Users/sacha/pkg/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /Users/sacha/pkg/julia/src/toplevel.c:636
jl_toplevel_eval_in at /Users/sacha/pkg/julia/src/builtins.c:606
eval at ./boot.jl:236
jlcall_eval_18559 at /Users/sacha/pkg/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/sacha/pkg/julia/src/gf.c:2196
eval_user_input at ./REPL.jl:66
unknown function (ip: 0x3198c9a76)
jl_apply_generic at /Users/sacha/pkg/julia/src/gf.c:2196
macro expansion at ./REPL.jl:97 [inlined]
#1 at ./event.jl:66
unknown function (ip: 0x3198c285f)
jl_apply_generic at /Users/sacha/pkg/julia/src/gf.c:2196
jl_apply at /Users/sacha/pkg/julia/src/./julia.h:1388 [inlined]
start_task at /Users/sacha/pkg/julia/src/task.c:261
Allocations: 7033163 (Pool: 7028643; Big: 4520); GC: 15
Segmentation fault: 11
```
Best!

### PR DESCRIPTION
**Pull request by tkelman**
miscompilation of broadcast

Will need to rebuild the windows binaries for appveyor before merging this, and the homebrew-julia llvm39 formula whenever. If you have any other LLVM patches in mind, do speak up so I can rebuild things fewer times.

fixes #19792, fixes #19769

### CODE REVIEW

**Comment by tkelman** [comment]
Note this doesn't address the likely test-system bug from https://github.com/JuliaLang/julia/issues/19792#issuecomment-269853074, so we will probably miss more of these going forward until we figure out what's up there.

File: deps/patches/llvm-3.9.0_win64-reloc-dwarf.patch [hunk 1]
```diff
@@ -1,8 +1,32 @@
+commit e345a27f2d4db7790fa8916969da7f84bf4decc7
+Author: Keno Fischer <kfischer@college.harvard.edu>
+Date:   Mon Jan 2 03:00:19 2017 +0000
+
+    Reapply "[CodeGen] Fix invalid DWARF info on Win64"
```
Comment (tkelman) [inline deps/patches/llvm-3.9.0_win64-reloc-dwarf.patch]: whoops this doesn't actually apply cleanly against 3.9.1

**Comment by nalimilan** [comment]
Patch D27389 does not seem to apply to 3.9.1 `--fuzz=0`. Here's an updated version that I'm using for the RPM package: https://gist.github.com/nalimilan/3c846e428c0d1f281c706ee26cd67c12

**Comment by ararslan** [comment]
LLVM 3.9 builds on FreeBSD on this branch, so I'd say this fixes #19769.

File: deps/patches/llvm-D28221-avx512.patch [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+     // Skylake:
+     case 0x4e:
+-      *Type = INTEL_COREI7; // "skylake-avx512"
+-      *Subtype = INTEL_COREI7_SKYLAKE_AVX512;
+-      break;
+     case 0x5e:
+-      *Type = INTEL_COREI7; // "skylake"
+-      *Subtype = INTEL_COREI7_SKYLAKE;
++      *Type = INTEL_COREI7;
++      if (Features & (1 << FEATURE_AVX512)) {
```
Comment (tkelman) [inline deps/patches/llvm-D28221-avx512.patch]: oops, looks like this had some review upstream and changed - will rebuild again

**Pull request merged**